Create NodeModelDAO, ClassLoader, a nodejs helper, and some i18n support.

mcarcaso opened the pull request.

mcarcaso committed:
Create the NodeModelFileDAO. A DAO for nodejs that only supports find.

mcarcaso committed:
Add awhile to async.

mcarcaso committed:
Add arequire to the context.

mcarcaso committed:
Create an OrDAO

mcarcaso committed:
Revert changes to Context.js

mcarcaso committed:
Introduce a ClassLoader that features an arequire method that will take a model ID and load it and its dependencies.

mcarcaso committed:
Create tools/node.js which can be used to execute a model from the command line via nodejs and add an example of its use.

mcarcaso committed:
Change the separator in NodeModelFileDAO to use require('path').sep.

mcarcaso committed:
Add Message support for i18n.

mcarcaso committed:
Make the error message in NodeModelFileDAO contain a little more info.

mcarcaso committed:
Add some extra error handling to the ClassLoader

mcarcaso committed:
Updates to node.js script. 1. Create the object in the context of the model that looked it up. 2. Add some error handling. 3. Add locale support.

mcarcaso committed:
Make use of the i18n support in the demo node script.

mcarcaso committed:
Remove a useless comma added to Context.js to make the merge touch one less file.

mcarcaso committed:
Update header years to be more accurate.

mcarcaso committed:
Add a TODO that mentions the current limitations of ClassLoader.

mcarcaso committed:
Make the node script do more error outputting.

mcarcaso committed:
Add an isRegistered method to Context.js

mcarcaso committed:
When creating the Requires, pass the subContext along.

mcarcaso committed:
Change the NodeModelFileDAO to return the model and to not do any lookups.

mcarcaso committed:
Modify the ClassLoader so that the model has its own arequire method and so each axiom can implement its own arequire. The ClassLoader's arequire now temporarily overrides foam.CLASS so it can get a handle on the model before it's registered.

mcarcaso committed:
Update the Test model in the node example so it makes use of an extend and require to show that arequire actually works.

adamvy-google commented on src/foam/core/ClassLoader.js:
```diff
@@ -17,48 +17,92 @@
+      }
+      for (var i = 0, a; a = this.axioms_[i]; i++) {
+        if (a.arequire) {
+          p = p.then(a.arequire());
```
I don't believe this is correct.  You can't pass a promise to p.then().

What you really want is to either build an array of promises

promises.push(a.arequire());

and then do return Promise.all(promises);

In fact I would restructure this whole method as such

var promises = [];

if ( this.extends ) promises.push(this.__subContext__.arequire(this.extends));

for ( var i = 0, a; a = this.axioms... ) {
  if ( a.arequire ) promises.push(a.arequire());
}

return Promise.all(promises);

mcarcaso replied:
Good call! Done.

adamvy-google commented on src/foam/core/ClassLoader.js:
```diff
@@ -17,48 +17,92 @@
-        // TODO: This method only adds requires. Make it add extends,
-        // properties, and anything else that might be needed.
+        if (this.__subContext__.isRegistered(modelId)) return Promise.resolve();
+        if (this.pending[modelId]) return this.pending[modelId];
```
I would test that this actually handles loops correctly.  I'm not sure it will, because by returning the previous promise you could end up in a situation where arequire('Foo') is waiting on arequire('Bar') which is waiting on the exisiting promise arequire('Foo');

I think you want to return an empty promise if pending is set.

mcarcaso replied:
I think this works ok but maybe I didn't run the right test. I created a circular dependency in the models located in demos/node/js and ran both the node script and the browser at http://localhost:8000/?model=nodetooldemo.Test&classpath=/src/,/demos/node/js/ and it seemed to be fine.

adamvy-google commented on src/foam/core/Requires.js:
```diff
@@ -88,10 +88,11 @@ foam.CLASS({
           var name  = a[1] || parts[parts.length-1];
-          return foam.core.Requires.create({name: name, path: path});
+          return foam.core.Requires.create(
+              {name: name, path: path}, this.__subContext__);
```
just pass `this` as the second parameter.  It will automatically pull out this.__subContext__ if it needs it.

mcarcaso replied:
Done

adamvy-google commented on src/foam/core/Requires.js:
```diff
@@ -88,10 +88,11 @@ foam.CLASS({
         }
 
-        return foam.core.Requires.create(o);
+        return foam.core.Requires.create(o, this.__subContext__);
```
ditto

mcarcaso replied:
Done

adamvy-google commented on src/foam/dao/NodeModelFileDAO.js:
```diff
@@ -29,17 +29,25 @@ foam.CLASS({
+      var model;
+      foam.CLASS = function(m) {
+        var cls = m.class ? foam.lookup(m.class) : foam.core.Model;
+        model = cls.create(m, X);
```
ditto, just pass this as second parameter, or in this case do var self = this; cls.create(m, self);

mcarcaso replied:
Done

adamvy-google commented on tools/node.js:
```diff
@@ -73,11 +73,13 @@ foam.CLASS({
-      });
+      var modelId = this.modelId;
+      return X.arequire(modelId)
+        .catch(console.log)
```
Did you test this?

I'm suprised you don't need to do console.log.bind(console)

kgrgreer replied:
They changed this in Chrome a couple releases ago, but this will break in browsers other than Chrome. Actually, I see that it works in Firefox also.

mcarcaso replied:
I think so. I can run the following line from the browser:
new Promise(function() { throw 'hello' }).catch(console.log);
and it just logs 'hello'

mcarcaso replied:
This script is only to be used when running node.js so I don't think we need to worry about how this works in browsers.

adamvy-google commented on tools/node.js:
```diff
@@ -73,11 +73,13 @@ foam.CLASS({
+      return X.arequire(modelId)
+        .catch(console.log)
+        .then(function() {
+          return foam.lookup(modelId).create(modelArgs, X).execute();
```
Again, use the parent object as a second parameter, not the context directly.

mcarcaso replied:
Done.

mcarcaso committed:
Make ClassLoader do a second isRegistered check before registering models because it's possible that the model was registered between starting the find and the find completing.

mcarcaso committed:
Add an index.html/index.js that can be used to easily load a detail view of a model from the browser.

mcarcaso committed:
Create an android string generator and a demo that makes use of it.

mcarcaso committed:
Create a circular requires in the node demo.

mcarcaso committed:
In index.js, create objects using self instead of X and force the view to showActions.

mcarcaso committed:
Change model's arequire to use an array of promises instead of chaining them via calls to then.

mcarcaso committed:
Create objects using 'this' rather than the subcontext.

mcarcaso committed:
Use this instead of X when creating models in the WebModelFileDAO.

mcarcaso committed:
Add the ability to specify classpaths when using index.html from browser. For example: http://localhost:8000/?model=nodetooldemo.Test&classpath=/src/,/demos/node/js/ (assuming webserver is running locally on port 8000)

adamvy-google commented on demos/node/js/nodetooldemo/TestCustomProperty.js:
```diff
@@ -0,0 +1,28 @@
+      name: 'value',
+      value: 'This is the custom property default value.',
+    },
+  ],
```
FOAM style is to not have any trailing commas.

adamvy-google commented on demos/node/run_script.sh:
```diff
@@ -0,0 +1,19 @@
+  modelId: 'nodetooldemo.Test',
+  modelArgs: {
+    name: 'Joe',
+  },
```
Trailing commas are technically illegal in JSON.  Although FOAM's json is a bit more permissive.

adamvy-google commented on index.js:
```diff
@@ -0,0 +1,107 @@
+      var modelId = this.model;
+      var viewId = this.view;
+      Promise.all([
+        X.arequire(viewId),
```
Can't we use this.arequire, since we imported it?

adamvy-google commented on demos/android/run_script.sh:
```diff
@@ -0,0 +1,42 @@
+rm $DIR/gen/*
+mkdir -p $DIR/gen
+
+read -r -d '' ARGS << EOM
```
Technically read -d is a bashism, so change the first line of these scripts to #!/bin/bash

adamvy-google commented on demos/android/run_script.sh:
```diff
@@ -0,0 +1,42 @@
+  locale: 'en',
+}
+EOM
+node $DIR/../../tools/node.js "$ARGS"
```
I'm also starting to think it might simplify things if we made these scripts just javascript scripts instead of shell scripts calling into foam.

You can do 

#!/usr/bin/env node

// JS code here
require('path/to/foam.js');

Something.create().execute();

etc.
etc.


Might be worth trying at some point to see if it simplifies things a bit.

adamvy-google replied:
We could even do some other fun things along this line, but I'll talk to you about that later.

adamvy-google commented on index.js:
```diff
@@ -0,0 +1,107 @@
+        X.arequire(viewId),
+        X.arequire(modelId),
+      ]).then(function() {
+        var model = X.lookup(modelId).create({}, self);
```
Don't pass the empty object as argument one.  It's faster to just pass null if you don't have any args to pass.

adamvy-google commented on index.js:
```diff
@@ -0,0 +1,107 @@
+        X.arequire(modelId),
+      ]).then(function() {
+        var model = X.lookup(modelId).create({}, self);
+        var view = X.lookup(viewId).create({showActions: true}, self);
```
If viewId was not specified we should do view = model.toE(null, self);

adamvy-google commented on index.js:
```diff
@@ -0,0 +1,107 @@
+  ],
+});
+
+var search = /([^&=]+)=?([^&]*)/g;
```
I kind of think that this code should be part of WebModelExecutor rather than top level code.

adamvy-google commented on src/foam/android/GenResources.js:
```diff
@@ -0,0 +1,72 @@
+      }
+      var p = Promise.resolve();
+      for (var i = 0; i < this.models.length; i++) {
+        p = p.then(this.arequire(this.models[i])) 
```
Same comments I made previously about building an array of promises and then doing Promise.all. I dont' think p.then(arequire()) does what you are expecting.

adamvy-google commented on src/foam/android/GenStrings.js:
```diff
@@ -0,0 +1,92 @@
+    }
+  ],
+  methods: [
+    function modelToResources(model) {
```
Might want to rename this to be classToResources() and name the parameter cls, since technically you're taking a Class not a Model.

adamvy-google commented on src/foam/core/ClassLoader.js:
```diff
@@ -0,0 +1,112 @@
+
+  methods: [
+    function arequire() {
+      var X = this.__subContext__;
```
I think we should use `this.__context__;`

Things you import should be taken from `__context__`, `__subContext__` is for things that we export.  And you are essentially importing arequire() here.

adamvy-google commented on src/foam/core/ClassLoader.js:
```diff
@@ -0,0 +1,112 @@
+
+  methods: [
+    function arequire() {
+      return this.__subContext__.arequire(this.path);
```
Again `this.__context__`

adamvy-google commented on src/foam/core/ClassLoader.js:
```diff
@@ -0,0 +1,112 @@
+      class: 'foam.core.ContextMethod',
+      code: function(X, modelId) {
+        if (X.isRegistered(modelId)) return Promise.resolve();
+        if (this.pending[modelId]) return this.pending[modelId];
```
Okay so this _would_ hang if you have a cycle, except that since you're registering the model before you call model.arequire(), this case is never hit via a circular dependency.

That being said, it feels more correct to me to do m.arequire() before we register the latch, as otherwise isRegistered() will return true for models that arn't quite loaded yet.  So I would make that change and then also fix this if to just return Promise.resolve()

adamvy-google commented on src/foam/dao/NodeModelFileDAO.js:
```diff
@@ -0,0 +1,54 @@
+    function find(id) {
+      var foamCLASS = foam.CLASS;
+      var self = this;
+      var X = this.__subContext__;
```
Is this variable used anywhere?

adamvy-google commented on src/foam/core/ClassLoader.js:
```diff
@@ -0,0 +1,112 @@
+      name: 'arequire',
+      class: 'foam.core.ContextMethod',
+      code: function(X, modelId) {
+        if (X.isRegistered(modelId)) return Promise.resolve();
```
FOAM style is to put spaces after the "(" and before the ")"

adamvy-google commented on src/foam/dao/NodeModelFileDAO.js:
```diff
@@ -0,0 +1,54 @@
+
+
+foam.CLASS({
+  package: 'foam.dao',
```
I'm not sure this it the right package for this.  Especially since its such an impoverished DAO.  Perhaps a foam.classloader package would be more appropriate.

I also think this should be added to files.js with a nodejs flag.

adamvy-google commented on src/foam/dao/WebModelFileDAO.js:
```diff
@@ -0,0 +1,71 @@
+
+
+foam.CLASS({
+  package: 'foam.dao',
```
Same comment as NodeModelFileDAO

adamvy-google commented on src/foam/dao/WebModelFileDAO.js:
```diff
@@ -0,0 +1,71 @@
+  methods: [
+    function find(id) {
+      var self = this;
+      var X = this.__subContext__;
```
Unused variable I assume?

adamvy-google commented on tools/node.js:
```diff
@@ -0,0 +1,89 @@
+      var modelArgs = this.modelArgs;
+      var X = this.__subContext__;
+      var modelId = this.modelId;
+      return X.arequire(modelId)
```
You imported arequire, so you can call it with just this.arequire(modelId)

adamvy-google reviewed (changes requested):
Leaving the i18n stuff for now, will review that later.

mcarcaso committed:
Remove some trailing commas.

mcarcaso committed:
Create an array of promises intead of chaining them with thens. Move most of the code in index.js inside of the model.

mcarcaso committed:
Make GenResources use an array of promises instead of chaining thens.

mcarcaso committed:
Make messages use a getter/setter so the value is correct when the locale changes.

mcarcaso committed:
Return the right promise in GenResources.

mcarcaso committed:
Replace bash scripts with js scripts.

mcarcaso committed:
Rename modelToResources to classToResources

mcarcaso committed:
Use __context__ instead of __subContext__ in Model's arequire

mcarcaso committed:
Use __context__ instead of __subContext__ for Requires arequire.

mcarcaso committed:
Use FOAM style for a couple of if statements that weren't correct.

mcarcaso committed:
Remove unused variable.

mcarcaso committed:
Remove unused variable.

mcarcaso committed:
Make the ClassLoader do an arequire on the model before registering it. Also keep track of what's arequired during an arequire to avoid circular dependencies.

mcarcaso committed:
Update node test to have two circular dependencies.

mcarcaso committed:
Move all of the classloader stuff into its own directory.

adamvy-google commented on demos/android/gen_xml.js:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env node
```
Look at that, 8 lines shorter! :D

mcarcaso commented:
Addressed most of your comments in your review individually but my comments seem to have been lost. In summary:
I've accounted for the circular dependency (if there's anything you want me to test, let me know)
I moved all of the classloader stuff into its own directory
I replaced bash scripts with js scripts

I'll remove the i18n stuff from this branch and re-introduce it on a different branch. I'll update this when that's all done.

adamvy-google commented on index.html:
```diff
@@ -1,8 +1,13 @@
+    <script src="src/foam/classloader/WebModelDisplayer.js"></script>
+    <script>
+      var executor = foam.classloader.WebModelDisplayer.create()
+      executor.fromQuery(this.window.location.search.substring(1));
```
Pretty sure in this context "this" and "window" are the same thing.

adamvy-google commented on src/foam/classloader/ClassLoader.js:
```diff
@@ -67,14 +67,25 @@ foam.CLASS({
-      code: function(X, modelId) {
-        if (X.isRegistered(modelId)) return Promise.resolve();
-        if (this.pending[modelId]) return this.pending[modelId];
+      code: function(X, modelId, deps) {
```
Maybe name it opt_deps

adamvy-google commented on src/foam/classloader/ClassLoader.js:
```diff
@@ -67,14 +67,25 @@ foam.CLASS({
           m.validate();
-          if (X.isRegistered(modelId)) return m;
+          return m.arequire(deps);
+        }).then(function() {
```
It might be useful to have model.arequire() resolve to the model.

Would make this slightly shorter as you could just get m as a parameter to this promise.

Not a big deal though, but something you could do if you wanted.

adamvy-google reviewed (commented):
LGTM overall.  Few small comments that you can choose to address at your discretion.

mcarcaso committed:
Remove everything that doesn't belong in the classloader stuff.

mcarcaso committed:
Remove a useless 'this'

mcarcaso committed:
Prepend opt_ to an optional arg

mcarcaso committed:
Return the model when arequiring it in ClassLoader's arequire method.

mcarcaso commented:
Thanks. Moved out the i18n stuff and addressed just about all of the comments.
